src/libraries/System.Private.Uri/src/System/Uri.cs

```diff
@@ -3262,16 +3262,17 @@ private unsafe void ParseRemaining()
             // For iri parsing if we found unicode the idx has offset into _originalUnicodeString..
             // so restart parsing from there and make _info.Offset.Path as _string.Length
 
-            idx = _info.Offset.Path;
-            origIdx = _info.Offset.Path;
+            idx = origIdx = _info.Offset.Path;
 
             //Some uris do not have a query
             //    When '?' is passed as delimiter, then it's special case
             //    so both '?' and '#' will work as delimiters
             if (buildIriStringFromPath)
             {
-                // Dos paths have no host.  Other schemes cleared/set _string with host information in PrivateParseMinimal.
-                if (IsDosPath)
+                DebugAssertInCtor();
```

Comment:
So this method can also be called from outside of ctor but not in `buildIriStringFromPath` case?
Also, why is the assert repeated in every `if (buildIriStringFromPath)`?

Reply:
Correct. Uri lazily parses the input in 3 steps (ctor, CreateUriInfo, ParseRemaining). The latter two are done lazily as needed.

In case the Uri contains non-ascii characters, a full parse will be triggered from the constructor, so the `buildIriStringFromPath` branch can only run in the ctor. It also has to, as it updates the `_string` field in a non-synchronized manner.

The `DebugAssertInCtor` method was added recently (#33042) to ensure thread safety assumptions hold, I just missed this specific part.

> Also, why is the assert repeated in every if (buildIriStringFromPath)

It's not strictly needed, I just prefer it over a comment emphasizing the same thing.

Reply:
Thanks for explanation.
> It's not strictly needed, I just prefer it over a comment emphasizing the same thing.
Perfectly fine with me.